### PR TITLE
Search Functionality

### LINKED ISSUE #10: Search
It should be great to search notebooks.

At the moment, we could search by filename.

Later, in a different issue after #3 is done, we could also search by title and/or content.

### PR DESCRIPTION
**Pull request by BrooklinJazz**
Co-authored-by: BAKER <pragmatic.inc@gmail.com>

- [X] Search by github filename
- [X] Refactored home page 
- [ ] Refactor many list_notebooks* functions into a single list_notebooks function with many options. 
- [ ] Mobile specific styles (Thinking the UI for search could be a bottom navbar or something nice on mobile 🤔) 
- [ ] Polish styles to ensure accuracy to previous version
- [X] Format project

image

I made some fairly opinionated changes in order to simplify the code.
- Random Button always displays with the same text
- Prefer list_notebooks function over list_notebooks_desc function (I think It makes sense to move to a single function with options, rather than many separate functions. I'm happy to chat about why if you want more context :)
- I simplified the `index.html.heex` template to make it easier to add the search bar without breaking styles. I reduced code size a fair bit, but I recommend you read through to ensure you are happy with the changes. Some styles are slightly "off" but I'm working on them.

 Open to feedback, but this PR is not fully ready to merge.

### CODE REVIEW

**Comment by BrooklinJazz** [comment]
I may have unintentionally changed the scrollbar.

However, I can't track down the change that caused it. The bar seems different locally than it does on the website even on the main branch  (for me) so maybe not 🤔

**Comment by hectorperez** [comment]
This looks great, Brooklin!

How about cherry-picking your commit `format project` — or even better run `mix format` again from `main` and open a new PR only with that?
Then, this "search" PR would have fewer changes and it would be easier to review.
I can do that if you wish. In fact, I should have done it already. Let me know if you'd like me to do this.

Regarding your comments:
- Having one `list_notebooks` with options is sensible 👌 
- Random button text sounds good.
- I run this locally and the styles look right. The scrollbar too. I'll review the changes after moving the changes from `mix format` out. Unless you think it is better to do it here.

**Comment by BrooklinJazz** [comment]
Sounds like a plan! I agree, it makes sense to have the mix format be a separate pr to make these changes cleaner. It's late for me so if you want to merge the format in that would be great, and I can fix any resulting issues tomorrow or whenever my next coding session is

**Comment by hectorperez** [comment]
I run `mix format` and merged it to `main`.

File: lib/notesclub_web/templates/page/index.html.heex [hunk 1]
```diff
@@ -1,58 +1,55 @@
...
+            <td>
+              <%= link "Run in Livebook", to: "https://livebook.dev/run?url=#{notebook.github_html_url}", target: "_blank" %>
+            </td>
+        </tbody>
+      <% end %>
+    </table>
+  </section>
+  <div class="text-center p-6">
+      <%= link "Random", to: Routes.index_path(@conn, :index), class: ["px-4 py-2 border border-gray-300 shadow-sm text-sm font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2"] %>
+      <%= link "Last week's", to: Routes.last_week_path(@conn, :last_week), class: [@page == :last_week && "hidden","px-4 py-2 border border-gray-300 shadow-sm text-sm font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2"] %>
```
Comment (hectorperez) [inline lib/notesclub_web/templates/page/index.html.heex:48]:
```suggestion
      <%= link "Last week", to: Routes.last_week_path(@conn, :last_week), class: [@page == :last_week && "hidden","px-4 py-2 border border-gray-300 shadow-sm text-sm font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2"] %>
```

Comment (hectorperez) [reply lib/notesclub_web/templates/page/index.html.heex:48]: Shall we shorten this button to `Last week`?

File: lib/notesclub/notebooks.ex [hunk 2]
```diff
@@ -17,14 +17,18 @@ defmodule Notesclub.Notebooks do
...
+      {:order, :desc}, query ->
+        order_by(query, [notebook], -notebook.id)
+
+      {:github_filename, github_filename}, query ->
+        search = "%#{github_filename}%"
+        where(query, [notebook], ilike(notebook.github_filename, ^search))
+
+      _, query ->
+        query
+    end)
```
Comment (hectorperez) [inline lib/notesclub/notebooks.ex:31]: That looks much better now 👏

**Review by hectorperez — APPROVED** [review]
This looks good!
The home page and `list_notebooks/1` are much better now, thanks!

Regarding styles, they look good to me. There is no need to match exactly what we had before. That was something quick I took from tailwindui.com

Regarding mobile styles, you could do it here or later in a new PR focused on making the site usable on a mobile. As you prefer ;)

Looking forward to seeing that search function in production.
Good job Brooklin!

Btw, after formatting the project in a different PR, I merged `main` into this PR and fixed conflicts. Let me know if I can help in any other way.

File: lib/notesclub_web/templates/page/index.html.heex (hunk 1 above)
Comment (BrooklinJazz) [reply lib/notesclub_web/templates/page/index.html.heex:48]: LGTM

File: lib/notesclub/notebooks.ex (hunk 2 above)
Comment (BrooklinJazz) [reply lib/notesclub/notebooks.ex:31]: 🥳🥳🥳🥳

**Comment by BrooklinJazz** [comment]
Thank you for fixing conflicts! I need to have a second read through this PR when I have some spare time, but it looks like it's close to being ready. Feel free to push it through if you 're happy with the changes and want to get this in quick, I just don't have much time RN until the weekend.

Thank you @hectorperez!

**Comment by hectorperez** [comment]
I merged it as @BigTom wanted to do #19 and it would have created more conflicts.

Thanks @BrooklinJazz!

**Pull request merged**